### PR TITLE
chore: add feed forward autoencoder

### PR DESCRIPTION
**Pull request by ChadElRicardo**
Created Python script and experiment notebook for Feed Forward Autoencoder

### CODE REVIEW

File: autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+__version__ = '1.0.0'
+__author__ = 'Richard Ricardo'
+
+import tensorflow as tf
+from tensorflow.keras.layers import Dense
+from keras.models import Model
```
Comment (AFAgarap) [inline autoencoder.py]: Why are we importing the `keras.models.Model` when we are using `tf.keras.Model`?

File: autoencoder.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+class Autoencoder(tf.keras.Model):
```
Comment (AFAgarap) [inline autoencoder.py:13]: Where's the experiment using this class? @ChadElRicardo

File: autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+def loss(reconstructed, original):
+    return tf.reduce_mean(tf.square(tf.subtract(reconstructed, original)))
+
+def train(loss, model, opt, original):
+    with tf.GradientTape() as tape:
+        reconstructed = model(original)
+        reconstruction_error = loss(reconstructed, original)
+    gradients = tape.gradient(reconstruction_error, model.trainable_variables)
+    gradient_variables = zip(gradients, model.trainable_variables)
+    opt.apply_gradients(gradient_variables)
```
Comment (AFAgarap) [inline autoencoder.py]: We can add a `tf.optimizers` object to the `Autoencoder` class instead of passing it as an argument.

Comment (AFAgarap) [reply autoencoder.py]: @ChadElRicardo In turn, the `learning_rate` will be a class parameter

Comment (AFAgarap) [reply autoencoder.py]: But of course, we can add a Boolean to the class if it's training or not. If it's not training, use a default `learning_rate`, so we can call the trained model for reconstructing images.

File: autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    return reconstruction_error
+
+def train_loop(model, loss, dataset, epochs):
```
Comment (AFAgarap) [inline autoencoder.py]: @ChadElRicardo We should add recording of images in TensorBoard, both the original and reconstructed, as well as recording of loss values.

**Review by AFAgarap — CHANGES_REQUESTED** [review]
@ChadElRicardo Kindly see the comments for improvement.

File: autoencoder.py (hunk 1 above)
Comment (ChadElRicardo) [reply autoencoder.py]: My bad, I just realize I don't need it. Thank you for this.

File: autoencoder.py (hunk 2 above)
Comment (ChadElRicardo) [reply autoencoder.py:13]: There's a second commit to this request **chore: add autoencoder notebook**  by the filename:  **feedfoward_autoencoder.ipynb** 

Thank you.

File: autoencoder.py (hunk 3 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Noted on this, will update the script accordingly. Thank you.

File: autoencoder.py (hunk 4 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Noted on this, will update the experiment notebook accordingly. Thank you.

File: autoencoder.py (hunk 3 above)
Comment (ChadElRicardo) [reply autoencoder.py]: @AFAgarap, is the Boolean also a class parameter?

File: autoencoder.py (hunk 2 above)
Comment (AFAgarap) [reply autoencoder.py:13]: I didn't see it. Sorry. I'll review it today. Thank you.

File: autoencoder.py (hunk 3 above)
Comment (AFAgarap) [reply autoencoder.py]: Yes.

**Review by AFAgarap — CHANGES_REQUESTED** [review]
Kindly share the Colab notebook with me, @ChadElRicardo So I can write my comments there. Thank you

File: autoencoder.py [hunk 5]
```diff
@@ -1,17 +1,32 @@
 """Implementation of feed forward autoencoder in TensorFlow 2.0"""
+
```
Comment (AFAgarap) [inline autoencoder.py]: Kindly remove this whitespace @ChadElRicardo

File: autoencoder.py [hunk 6]
```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
...
+                tf.summary.scalar('loss', train_loss.result(),
+                                  step=epoch)
+            epoch_loss += loss_values
+        model.loss.append(tf.reduce_mean(epoch_loss))
+        print('Epoch {}/{}. Loss: {}'.format(epoch + 1, epochs,
+              tf.reduce_mean(epoch_loss)))
+        train_loss.reset_states()
+
+
+def log_images(test_features, model):
```
Comment (AFAgarap) [inline autoencoder.py]: Let's move this to a different module, say, `utils.py` @ChadElRicardo

File: autoencoder.py [hunk 7]
```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
...
+        images = np.reshape(originals, (-1, 28, 28, 1))
+        tf.summary.image('Original', images, max_outputs=10, step=0)
+
+    with file_writer.as_default():
+        images = np.reshape(reconstructed, (-1, 28, 28, 1))
+        tf.summary.image('Reconstructed', images, max_outputs=10,
+                         step=0)
+
+
+train_loss = tf.keras.metrics.Mean('train_loss', dtype=tf.float32)
```
Comment (AFAgarap) [inline autoencoder.py]: Why do we have these lines? [165-168]
Kindly use [this notebook](https://github.com/AFAgarap/autoencoders/blob/master/notebooks/0-autoencoder.ipynb) as a template.

File: autoencoder.py [hunk 8]
```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
...
-  for epoch in range(epochs):
-      epoch_loss = 0
-      for batch_features in dataset:
-          loss_values = train(loss, model, opt, batch_features)
-          epoch_loss += loss_values
-      model.loss.append(tf.reduce_mean(epoch_loss))
-      print('Epoch {}/{}. Loss: {}'.format(epoch + 1, epochs, tf.reduce_mean(epoch_loss)))
\ No newline at end of file
+
+def train_loop(
```
Comment (AFAgarap) [inline autoencoder.py]: Let's also write this down to `utils.py`

Comment (AFAgarap) [reply autoencoder.py]: Also, `train` would be enough as a function name

File: autoencoder.py [hunk 9]
```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
...
+        return self.optimizer
+
 
 def loss(reconstructed, original):
-    return tf.reduce_mean(tf.square(tf.subtract(reconstructed, original)))
+    return tf.reduce_mean(tf.square(tf.subtract(reconstructed,
+                          original)))
+
+
+def train(
```
Comment (AFAgarap) [inline autoencoder.py]: Let's use `train_step` as a function name

File: autoencoder.py [hunk 10]
```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
...
+                    self.decoder_hidden_layers[layer](activation)
+        return self.decoder_output_layer(activation)
+
+    def opt(self):
+        return self.optimizer
+
 
 def loss(reconstructed, original):
-    return tf.reduce_mean(tf.square(tf.subtract(reconstructed, original)))
+    return tf.reduce_mean(tf.square(tf.subtract(reconstructed,
```
Comment (AFAgarap) [inline autoencoder.py]: For better readability, how about --
```
reconstruction_error = tf.square(tf.subtract(reconstructed, original))
reconstruction_error = tf.reduce_mean(reconstruction_error)
```

File: autoencoder.py [hunk 11]
```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
...
-      if self.num_hidden_layers > 1:
-        for layer in range(self.num_hidden_layers - 1):
-          layer = layer + 1
-          activation = self.decoder_hidden_layers[layer](activation)
-      return self.decoder_output_layer(activation)
+        """Encoder"""
+
+        activation = self.encoder_hidden_layers[0](input_features)
+        if self.num_hidden_layers > 1:
+            for layer in range(self.num_hidden_layers - 1):
```
Comment (AFAgarap) [inline autoencoder.py]: We can simply use it as it is, `range(self.num_hidden_layers)`.
We start with index 0 anyway (supposedly).

File: autoencoder.py [hunk 12]
```diff
@@ -1,17 +1,32 @@
...
 import tensorflow as tf
+import datetime
 from tensorflow.keras.layers import Dense
-from keras.models import Model
+
 
 class Autoencoder(tf.keras.Model):
-    def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers, encoder_decrements, decoder_increments):
+
+    def __init__(
```
Comment (AFAgarap) [inline autoencoder.py]: I think we can simplify our class by passing a parameter like `neurons=[512, 512]`. That is, we can do something like

```
neurons = [512, 512]
model = AE(neurons=neurons, num_layers=len(neurons), ...)
```

File: autoencoder.py [hunk 13]
```diff
@@ -1,17 +1,32 @@
...
+        self,
+        intermediate_dim,
+        code_dim,
+        original_dim,
+        num_hidden_layers,
+        encoder_decrements,
+        decoder_increments,
+        learning_rate=1e-3,
```
Comment (AFAgarap) [inline autoencoder.py]: Sorry, upon thinking about it, we can remove this, and maintain the optimizer outside the class.

**Review by AFAgarap — CHANGES_REQUESTED** [review]
Hello, @ChadElRicardo Kindly review the comments I wrote. Thank you.

File: autoencoder.py (hunk 8 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Hi @AFAgarap, is the **utils.py** will be containing both `train` & `train_step` function?

File: autoencoder.py (hunk 6 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Hi @AFAgarap, is the **utils.py** will be containing both `train` & `train_step` function?

File: autoencoder.py (hunk 7 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Lines 165-168 are for the TensorBoard since the logging of the **training performance**, **original** & **reconstructed images** that are within the module. I think I asked if the TensorBoard logging should be in the module, and I assumed you stated that would be fine.

File: autoencoder.py (hunk 9 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Noted on this. Thank you.

File: autoencoder.py (hunk 10 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Noted on this. Thank you.

File: autoencoder.py (hunk 11 above)
Comment (ChadElRicardo) [reply autoencoder.py]: Noted on this. Thank you.

File: autoencoder.py (hunk 6 above)
Comment (AFAgarap) [reply autoencoder.py]: Yes. They can be moved there, so we can reuse them.

File: autoencoder.py (hunk 7 above)
Comment (AFAgarap) [reply autoencoder.py]: Oh, sorry. My bad. Let's move it to the notebook. The module itself should only consist of the class, while the helper module would have the helper functions like `train` and `train_step`.

File: autoencoder.py (hunk 8 above)
Comment (AFAgarap) [reply autoencoder.py]: Answered above.

**Comment by AFAgarap** [comment]
@ChadElRicardo I'm deleting this branch. Please make a new branch for other modules you will write.
